### ISSUE TITLE
[Feature Request]: Potential Next.js support options

### ISSUE THREAD

**Issue by zackarychapple** [web-infra-dev/rspack#2407]
### What problem does this feature solve?

There are a lot of people who would be interested in using Next.js with Rspack. Today @ScriptedAlchemy and I were digging into it.  @valorkin pointed out to us that we can use `NEXT_PRIVATE_LOCAL_WEBPACK` to replace the webpack config. Additionally we can alias webpack to Rspack via the package manager. This gets us to the object.assign that is being done in the [webpack.js](https://github.com/vercel/next.js/blob/75c6b100090c2cb1854ef888257ff8348a2df956/packages/next/src/compiled/webpack/webpack.js#L6). It seems that if we can expose the bindings and resolve these assigns to Rspack we may be able to get Rspack building Next.js. 

```      
BasicEvaluatedExpression: require('webpack/lib/javascript/BasicEvaluatedExpression'),
ModuleFilenameHelpers: require('webpack/lib/ModuleFilenameHelpers'),
NodeTargetPlugin: require('webpack/lib/node/NodeTargetPlugin'),
StringXor: require('webpack/lib/util/StringXor'),
NormalModule: require('webpack/lib/NormalModule'),
sources: require('webpack').sources,
webpack: require('webpack')
```

Any thoughts @hardfist?

### What does the proposed API of configuration look like?

Surface the above hooks and internal plugins as a binding that can be found on the root Rspack object or another interface that is specified and accessible.

**Comment by zackarychapple** [comment]
> There is likely some additional internal plugins/hooks next would use - but these bindings would be the initial requirement to better understand whats needed/missing

Correct, this would get us past the initial crashing to see what individual hooks are being used.

**Comment by hardfist** [comment]
I actually did try port nextjs on rspack with @ijjk and I don't think it's possible to port nextjs on rspack without modifying lots of nextjs code, and since nextjs are betting on turbopack right now i think it's low priority to run next.js on rspack right now, we may revisit this issue after rspack supports more and more hooks & api.

**Comment by ScriptedAlchemy** [comment]
As workaround, we can call a childCompiler in next webpack & disable parser of a large segment of the codebase, then one could hook into additionalAssets to pass it through next/webpack haha.

**Comment by zmzlois** [comment]
> As workaround, we can call a childCompiler in next webpack & disable parser of a large segment of the codebase, then one could hook into additionalAssets to pass it through next/webpack haha.

I was looking into how to build a monorepo includes three nextjs apps using rspack today following a conference talk in [CityJS London](https://london.cityjsconf.org/) delivered by @ScriptedAlchemy 

But didn't find any migration guide out of the box, wondering what a workaround for this as @ScriptedAlchemy  mentioned would look like? Is Rspack by default focusing on supporting developers based in China and must use modern.js?

**Comment by zmzlois** [comment]
As of 3rd April, 2023, create-react-app has been removed from the official react documentation after [this pr](https://github.com/reactjs/react.dev/pull/5487) and the official recommendation is default to
``` 
create-next-app

create-gatsby

create-expo-app
```

seen [here](https://react.dev/learn/start-a-new-react-project)

**Comment by ScriptedAlchemy** [comment]
Gatsby has a beta with rspack, so that's one down! 

Next is harder and controlled by vercel. We can't change next.js source. Other projects want to use rspack. 

I do think that there is a niche application for this exact type of problem though. For instance, supporting module federation in next. 
As a workaround, I'm drawing on a similar strategy I used in next. 

Imagine a RspackWebpackPlugin() - this plugin would disable webpack parser, effectively turning off code processing. 

Webpack would pass it's options to rspack and rspack would bind all its hooks back to webpacks api. So webpack itself wouldn't do much work (the slow parts) I'd try to make webpack emit the work of rspack. 

It's a huge hack. But might work for stubborn cases, like next. 

Expo is react native I think? We have nativescript I think, so rspack might have enough hooks available to support a react native build. 
To support react native, we would use callstack/repack plugin and see if it works with rspack. Expo is it's own build tool - but I get what you're pointing out. Market penetration on these big frameworks. 

I'll create a issue for repack investigation.

**Comment by zackarychapple** [comment]
> We have nativescript I think, so rspack might have enough hooks available to support a react native build.

There is a few hooks needed for NativeScript, however I believe some have landed. Maybe @edusperoni can check again after we land everything for Angular. They were blocked by similar hooks.

**Comment by jkzing** [comment]
> Is Rspack by default focusing on supporting developers based in China and must use modern.js?

@zmzlois No, that's not accurate. Modern.js is simply a framework that utilizes Rspack as its bundler, like what Gatsby does. It's important to note that both Modern.js and Rspack are global projects and not only focusing on supporting China developers.
